### DISCUSSION TITLE
v4.0.1

### DISCUSSION THREAD

**Discussion by HolyWu** [HolyWu/vs-realesrgan#17] [Announcements]
- Switch to PyTorch again for inference. A few parameters are added and some parameters are removed.
- Add official ESRGAN x4 model and realesr-general-x4v3 model.

See [Discussions](https://github.com/HolyWu/vs-realesrgan/discussions) for benchmarks of some models.

<hr /><em>This discussion was created from the release <a href='https://github.com/HolyWu/vs-realesrgan/releases/tag/v4.0.1'>v4.0.1</a>.</em>

**Comment by HolyWu** [comment]
# Benchmark

Configuration: NVIDIA RTX 3050, driver 527.37, Windows 10 21H2, Python 3.10.8, VS R61, 720x480 FP16
Data format: FPS / VRAM usage

## RealESRGAN_x2plus

| Args                                          | Result                 |
| --------------------------------------------- | ---------------------- |
| num_streams=1                                 | 2.69 fps /   832,244 K |
| num_streams=2                                 | 2.85 fps / 1,446,644 K |
| num_streams=1, nvfuser=True, cuda_graphs=True | 2.82 fps /   860,908 K |
| num_streams=2, nvfuser=True, cuda_graphs=True | 2.89 fps / 1,555,192 K |
| num_streams=1, trt=True                       | 4.52 fps /   596,708 K |
| num_streams=2, trt=True                       | 4.74 fps / 1,090,284 K |

## RealESRGAN_x4plus

| Args                                          | Result                 |
| --------------------------------------------- | ---------------------- |
| num_streams=1                                 | 0.68 fps / 2,587,388 K |
| num_streams=2                                 | 0.70 fps / 4,956,944 K |
| num_streams=1, nvfuser=True, cuda_graphs=True | 0.71 fps / 2,497,272 K |
| num_streams=2, nvfuser=True, cuda_graphs=True | 0.73 fps / 4,827,924 K |
| num_streams=1, trt=True                       | 1.18 fps / 1,942,252 K |
| num_streams=2, trt=True                       | 1.21 fps / 3,783,424 K |

## RealESRGAN_x4plus_anime_6B

| Args                                          | Result                 |
| --------------------------------------------- | ---------------------- |
| num_streams=1                                 | 2.00 fps / 2,562,812 K |
| num_streams=2                                 | 2.12 fps / 4,932,368 K |
| num_streams=1, nvfuser=True, cuda_graphs=True | 2.07 fps / 2,409,204 K |
| num_streams=2, nvfuser=True, cuda_graphs=True | 2.24 fps / 4,674,312 K |
| num_streams=1, trt=True                       | 3.58 fps / 1,876,716 K |
| num_streams=2, trt=True                       | 3.81 fps / 3,652,344 K |

## realesr-animevideov3

| Args                                          | Result                |
| --------------------------------------------- | --------------------- |
| num_streams=1                                 |  9.68 fps / 387,820 K |
| num_streams=2                                 | 13.01 fps / 590,576 K |
| num_streams=1, nvfuser=True, cuda_graphs=True | 14.41 fps / 287,460 K |
| num_streams=2, nvfuser=True, cuda_graphs=True | 23.45 fps / 439,012 K |
| num_streams=1, trt=True                       | 14.20 fps / 361,184 K |
| num_streams=2, trt=True                       | 19.97 fps / 621,284 K |

**Reply by dan64** [reply]
I noted that the version 4.0.1 dropped the support to ONNX. Previously I was using the version 3.0.0 with ONNX support. Upgrading to version 4.0.1 I noted a big improvement on the start-up time when the TensorRT is enabled. Previously it was necessary to wait till 40min before the encoding could start and was generated a cache file with name like: "TensorrtExecutionProvider_TRTKernel_graph_torch-jit-export_18011103394555784438_1_0_fp16.engine". Now the start-up with TensorRT is less than 1min and are not generated the cache files, the cache is in memory ?.
It seems that this improvement in using TensorRT has impacted on the encoding speed. Now, on my RTX 3060, the encoding using "x4plus_anime_6B" (old name) is 15% slower (2.62fps vs 3.08fps) and the encoding using animevideo_v3 is 27% slower respect to animevideo_v2 (13.58fps vs 18.70fps) with no observable difference in quality.
Could this change in speed due to the missing support to ONNX ? If this is the case it is possible to realese a version with ONNX ?

Thank you,
Dan

**Reply by HolyWu** [reply]
The engine file is still cached in the vsrealesrgan directory under Python's site-packages by default after it is successfully built. You should see some logs in the console when the engine file is built for the first time.

**Reply by dan64** [reply]
To be sure to see the cache files, I used the following command (in Windows):
clip = **RealESRGAN**(clip=clip, model=3, device_index=0, num_streams=2, trt_cache_path="C:/Temp", trt_max_workspace_size=1073741824)
but cache files are not created.

**Reply by HolyWu** [reply]
Did you specify `trt=True`? trt is False by default.

**Reply by dan64** [reply]
You are right. I forgot to set "trt=True". Now the cache file is generated and also the speed (with num_streams=2) is good. Respect to the version 3.0.0 the speed using animevideo_v3 increased by about 20% (from 18.70fps to 22.40fps).
Thanks!

**Reply by HolyWu** [reply]
For unknown reason `nvfuser=True, cuda_graphs=True` is faster than TensorRT on animevideov3 for me. I suggest you to try it too.

**Reply by dan64** [reply]
Thank you for the suggestion. I noted that by increasing "num_streams" above 2 the gain in speed is low. Do you know for a GPU with 12GB what is its optimal value ?

**Reply by HolyWu** [reply]
That's expected. Most of the time num_streams=2 or 3 already saturates consumer GPU. Increasing num_streams further causes more workloads to idle/wait on the queue with no significant speed gain.

**Reply by dan64** [reply]
On my RTX 3060 with 12GB, I obtained the following encoding speed for **realesr-animevideov3**

num_streams=1,  17.64fps
num_streams=2,  21.46fps (+21.6%)
num_streams=3,  22.41fps (+27.0%)
num_streams=4,  24.11fps (+36.7%)
num_streams=5,  23.76fps (+34.7%)

so in this case it seems that num_streams=4 is the optimal value, but I'm not sure if this result can be generalized for any input resolution and model.

**Reply by dan64** [reply]
Using **RealESRGAN_x4plus_anime_6B** (my preferred model) I found that the optimal value is num_streams=3, probably this value represent the safest value that can be used with 12GB.

**Reply by dan64** [reply]
Hello HolyWu, I'd like to know if you have a plan to add **x2** model for **RealESRGAN_x4plus_anime_6B**. It could be useful to speed up the encoding when a 2x multiplier is enough (for example a DVD source).

**Reply by HolyWu** [reply]
I can't, because the author of [Real-ESRGAN](https://github.com/xinntao/Real-ESRGAN) only published x4 model of [RealESRGAN_x4plus_anime_6B](https://github.com/xinntao/Real-ESRGAN/blob/master/docs/model_zoo.md#for-anime-images--illustrations).

**Reply by dan64** [reply]
Hello HolyWu, the author changed the model to support also 2x, see the answer at the following link: [https://github.com/xinntao/Real-ESRGAN/issues/523#issuecomment-1367535597](url)

**Reply by HolyWu** [reply]
The guy replied to you (TechnoMasterBoy) is not the author. The author is xinntao. I bet what TechnoMasterBoy did is simply upscaling to 4x first and then downscaling to 2x afterwards.
